File: Leetcode/Median of Two Sorted Arrays/Median of two array.java

/*Java program to find median of two sorted arrays*/
import java.io.*;
import java.util.*;
class Median {
    static double findMedianSortedArrays(int[] nums1, int[] nums2) {
        
        //Find the mid 
        int mid = ((nums1.length + nums2.length) / 2) + 1;
        int[] result = new int[mid];
        int i1 = 0, i2 = 0, ir=0;
        while (ir < mid && i1 < nums1.length && i2 < nums2.length){
            int val1 = nums1[i1];
            int val2 = nums2[i2];
            if(val1 < val2){
                result[ir++] = val1;
                i1++;
            }else{
                result[ir++] = val2;
                i2++;
            }
        }

        int il = i1 < nums1.length ? i1 : i2;
        int[] numl = i1 < nums1.length ? nums1 : nums2;
        for (; ir < result.length; il++){
            result[ir++] = numl[il];
        }
        // If the length of array is even
        if ((nums1.length + nums2.length) % 2 == 0) {
            return (double) ((double) result[mid-2] + (double) result[mid-1]) / (double) 2;
        }
        // If the length of array is odd
         else {
            return (double) result[mid-1];
        }
    }
    //Driver's code
    public static void main(String args[]){
        Scanner sc = new Scanner(System.in);
        int n=sc.nextInt();
        int m=sc.nextInt();
        int a1[] = new int[n];
        int a2[] = new int[m];
        for(int i=0;i<n;i++)
          a1[i]=sc.nextInt();
        for(int i=0;i<m;i++)
          a2[i]=sc.nextInt();

    double ans = findMedianSortedArrays(a1,a2);
    System.out.println(ans);
    }
}

File: Leetcode/Median of Two Sorted Arrays/Median of two array.c

// C Program to find median of two sorted Array
double findMedianSortedArrays(int* nums1, int nums1Size, int* nums2, int nums2Size){

    //num3Size length
int nums3Size = nums1Size+nums2Size;
    int *nums3 = (int *) malloc(sizeof(int)*nums3Size);
    int j=0, k=0;
    //if num3size is zero return zero
    if (nums3Size == 0) 
        return 0;
        //if nums1Size is zero
    else if (nums1Size == 0) {
        for (int i=0; i < nums2Size; i++) {
            nums3[i] = nums2[i];
        }
    }

    //if nums2Size is zero
    else if (nums2Size == 0) {
        for (int i=0; i < nums1Size; i++) {
            nums3[i] = nums1[i];
        }
    }
    ////if nums1Size, nums2Size and nums3Size are not zero
    else {
        for (int i=0; i < nums3Size; i++) {
            if (k == nums2Size && j < nums1Size){
                nums3[i] = nums1[j];
                j++;
            }
            else if (j == nums1Size && k < nums2Size) {
                nums3[i] = nums2[k];
                k++;
            }
            else if (nums2[k] < nums1[j]) {
                nums3[i] = nums2[k];
                k++;
            }
            else if (nums1[j] <= nums2[k]){
                nums3[i] = nums1[j];
                j++;
            }
        }
    }

    //if nums3Size is 1
    if (nums3Size == 1)
        return nums3[0];
        //checking if nums3Size is even and return answer
   else if (nums3Size % 2 == 0) {
       int index = nums3Size/2;
       double median = (nums3[index-1]+nums3[index])/(double) 2;
       return median;
   }
   //checking is nums3Size is odd and return answer
    else {
        printf("odd\n");
        return nums3[nums3Size/2];
    }
    return -1;
}

